From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Thu, 7 Nov 2024 21:50:47 +0100
Subject: [PATCH] Lithium: fast util

This patch is based on the following mixins:
* "net/caffeinemc/mods/lithium/mixin/math/fast_util/AABBMixin.java"
* "net/caffeinemc/mods/lithium/mixin/math/fast_util/AxisCycleDirectionMixin.java"
* "net/caffeinemc/mods/lithium/mixin/math/fast_util/DirectionMixin.java"
By: 2No2Name <2No2Name@web.de>
As part of: Lithium (https://github.com/CaffeineMC/lithium)
Licensed under: LGPL-3.0 (https://www.gnu.org/licenses/lgpl-3.0.html)

diff --git a/net/minecraft/core/AxisCycle.java b/net/minecraft/core/AxisCycle.java
index 0dd0313e507614109abbd469de7eecda9f6d55b0..0ba48a086f7677939aa96c1c62fb6d3e9dacfaef 100644
--- a/net/minecraft/core/AxisCycle.java
+++ b/net/minecraft/core/AxisCycle.java
@@ -35,7 +35,17 @@ public enum AxisCycle {
 
         @Override
         public Direction.Axis cycle(Direction.Axis axis) {
-            return AXIS_VALUES[Math.floorMod(axis.ordinal() + 1, 3)];
+            // Leaf start - Lithium - fast util
+            return switch (axis.ordinal()) {
+                case 0 -> //X
+                    Direction.Axis.Y;
+                case 1 -> //Y
+                    Direction.Axis.Z;
+                case 2 -> //Z
+                    Direction.Axis.X;
+                default -> throw new IllegalArgumentException();
+            };
+            // Leaf end - Lithium - fast util
         }
 
         @Override
@@ -56,7 +66,17 @@ public enum AxisCycle {
 
         @Override
         public Direction.Axis cycle(Direction.Axis axis) {
-            return AXIS_VALUES[Math.floorMod(axis.ordinal() - 1, 3)];
+            // Leaf start - Lithium - fast util
+            return switch (axis.ordinal()) {
+                case 0 -> //X
+                    Direction.Axis.Z;
+                case 1 -> //Y
+                    Direction.Axis.X;
+                case 2 -> //Z
+                    Direction.Axis.Y;
+                default -> throw new IllegalArgumentException();
+            };
+            // Leaf end - Lithium - fast util
         }
 
         @Override
@@ -65,6 +85,15 @@ public enum AxisCycle {
         }
     };
 
+    // Leaf start - Lithium - fast util
+    static {
+        assert Direction.Axis.X.ordinal() == 0;
+        assert Direction.Axis.Y.ordinal() == 1;
+        assert Direction.Axis.Z.ordinal() == 2;
+        assert Direction.Axis.values().length == 3;
+    }
+    // Leaf end - Lithium - fast util
+
     public static final Direction.Axis[] AXIS_VALUES = Direction.Axis.values();
     public static final AxisCycle[] VALUES = values();
 
diff --git a/net/minecraft/core/Direction.java b/net/minecraft/core/Direction.java
index 0206ca3d71ec2aefd9c22ad7e828f565f6b4528c..c045411879e2993bea08ad8c45c44dbbe0dcfb5c 100644
--- a/net/minecraft/core/Direction.java
+++ b/net/minecraft/core/Direction.java
@@ -225,7 +225,7 @@ public enum Direction implements StringRepresentable, ca.spottedleaf.moonrise.pa
     }
 
     public Direction getOpposite() {
-        return this.opposite; // Paper - optimise collisions
+        return VALUES[this.oppositeIndex]; // Leaf - Lithium - fast util
     }
 
     public Direction getClockWise(Direction.Axis axis) {
@@ -357,7 +357,7 @@ public enum Direction implements StringRepresentable, ca.spottedleaf.moonrise.pa
     }
 
     public static Direction getRandom(RandomSource random) {
-        return Util.getRandom(VALUES, random);
+        return VALUES[random.nextInt(VALUES.length)]; // Leaf - Lithium - fast util
     }
 
     public static Direction getApproximateNearest(double x, double y, double z) {
diff --git a/net/minecraft/world/phys/AABB.java b/net/minecraft/world/phys/AABB.java
index 120e432df71eaee94225555d24c82e6d05b34b9f..7494a012959660e31b1e6d63b27ca85863be44f8 100644
--- a/net/minecraft/world/phys/AABB.java
+++ b/net/minecraft/world/phys/AABB.java
@@ -19,6 +19,15 @@ public class AABB {
     public final double maxY;
     public final double maxZ;
 
+    // Leaf start - Lithium - fast util
+    static {
+        assert Direction.Axis.X.ordinal() == 0;
+        assert Direction.Axis.Y.ordinal() == 1;
+        assert Direction.Axis.Z.ordinal() == 2;
+        assert Direction.Axis.values().length == 3;
+    }
+    // Leaf end - Lithium - fast util
+
     public AABB(double x1, double y1, double z1, double x2, double y2, double z2) {
         this.minX = Math.min(x1, x2);
         this.minY = Math.min(y1, y2);
@@ -80,11 +89,32 @@ public class AABB {
     }
 
     public double min(Direction.Axis axis) {
-        return axis.choose(this.minX, this.minY, this.minZ);
+        // Leaf start - Lithium - fast util
+        return switch (axis.ordinal()) {
+            case 0 -> //X
+                this.minX;
+            case 1 -> //Y
+                this.minY;
+            case 2 -> //Z
+                this.minZ;
+            default -> throw new IllegalArgumentException();
+        };
+
+        // Leaf end - Lithium - fast util
     }
 
     public double max(Direction.Axis axis) {
-        return axis.choose(this.maxX, this.maxY, this.maxZ);
+        // Leaf start - Lithium - fast util
+        return switch (axis.ordinal()) {
+            case 0 -> //X
+                this.maxX;
+            case 1 -> //Y
+                this.maxY;
+            case 2 -> //Z
+                this.maxZ;
+            default -> throw new IllegalArgumentException();
+        };
+        // Leaf end - Lithium - fast util
     }
 
     @Override
